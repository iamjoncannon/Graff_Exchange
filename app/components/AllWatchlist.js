import React from 'react';
import {isDesktop} from './utils'
import AddSymbolBox from "./AddSymbolBox"

const dummyData = [
  ["Bitcoin", "BTC", "4878", "2.54"],
  ["Bitcoin4", "BTC", "4878", "2.54"],
  ["Bitcoinr", "BTC", "4878", "2.54"],
  ["Bitcoin", "BTC", "4878", "2.54"],
  ["Bitcoin", "BTC", "4878", "2.54"],
  ["Bitcoin", "BTC", "4878", "2.54"],
  ["Bitcoin", "BTC", "4878", "2.54"],
  ["Bitcoin", "BTC", "4878", "2.54"],
  ["Bitcoin", "BTC", "4878", "2.54"],
  ["Bitcoin", "BTC", "4878", "2.54"],
]

export default class AllWatchList extends React.Component {
 
  constructor(props) {
    super(props);

    this.state = {

        isModalShowing : false,
    }
   }

    closeModal = () => {

      this.setState({ isModalShowing : false })
    }
  
    render() {

      const {isModalShowing, whichModal} = this.state

 
      return (
        <div>

        <div className="all-watchlist">
        
          <div>
            <i 
              className={`fas fa-plus fa-${isDesktop() ? "2" : "7"}x`} 
              onClick={()=>this.setState({isModalShowing: true})}
            />
          
            <i className={`fas fa-edit fa-${isDesktop() ? "2" : "7"}x`} />
          
          </div>

          <span>Portfolio</span>

          {
            dummyData.map((item, i)=>{

              return(

                <div>
                  <div>

                    <span>{item[0]}</span>
                    <span>{item[1]} </span>
                    
                  </div>

                  <div>
              
                    <span>{item[2]}</span>
                    <span>{item[3]}</span>
                  </div>
                </div>
              )
            })
          }
        </div>

          {
            isModalShowing && 

            <div className="modal-container" onClick={()=>this.setState({isModalShowing: false})}>

              <div onClick={ (e)=> e.stopPropagation()} >
                
                    <AddSymbolBox />
                  

              </div>

            </div>
          }

        </div>
      );
    };
}
