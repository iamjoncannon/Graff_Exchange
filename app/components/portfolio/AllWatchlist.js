import React from 'react';
import { connect } from "react-redux";
import { isDesktop } from '../utils'
import AddSymbolBox from "./AddSymbolBox"

class AllWatchList extends React.Component {
 
  constructor(props) {
    super(props);

      this.state = {

          isModalShowing : false,
          edit: false,
          hideList: []
      }
   }

   componentDidMount(){

      let storedHideList = localStorage.getItem('hideList')

      this.setState({

        hideList: storedHideList ? JSON.parse( localStorage.getItem('hideList') ) : [] 
      })
   }

    closeModal = () => {

      this.setState({ isModalShowing : false })
    }

    handleEdit = (item) =>{
      
      const { hideList } = this.state
      let { symbol } = item[1]
      
      let newHidelist 

      if(hideList.includes(symbol)){

        newHidelist = hideList.filter(item => item !== symbol)
      }
      else{

        newHidelist = [ ...this.state.hideList, item[1].symbol ]
      }

      localStorage.setItem('hideList', JSON.stringify(newHidelist))

      this.setState({

        hideList : newHidelist
      })
    }
  
    render() {

      const { isModalShowing, edit } = this.state
      const { portfolio, token } = this.props
  
      return (

        <div>

        <div className="all-watchlist">
        
          <div>

            <i 
              className={`fas fa-plus fa-plus-edit fa-${isDesktop() ? "2" : "7"}x`} 
              onClick={()=>this.setState({edit: false, isModalShowing: true})}
            />
          
            <i 
              className={`fas fa-edit fa-${isDesktop() ? "2" : "7"}x`} 
              onClick={()=>this.setState({edit: !edit, isModalShowing: false})}            
            />
          
          </div>

          <span>Portfolio</span>
          
          <div>

            { portfolio &&

              Object.entries(portfolio)
                .sort()
                .filter((item)=>{

                    if(this.state.edit){

                      return true
                    }
                    else{

                      return !this.state.hideList.includes(item[1].symbol)
                    }
                })
                .map((item, i)=>{

                  const { data, symbol } = item[1]

                  const { hideList } = this.state 

                  return(

                  <div key={i} className="watchlist-item-container">

                      { this.state.edit && 
                        

                        <i 
                          className={`fas fa-${ hideList.includes(symbol) ? "plus" : "minus"} fa-${isDesktop() ? "2" : "7"}x`} 
                          onClick={()=> this.handleEdit(item) }
                        />
                      
                      }

                    { data &&
                      <div className="watchlist-item" style={{opacity: hideList.includes(symbol) ? ".5" : "1" }}>


                          <div>

                            <span style={{ color : data.latestPrice < item[1].price ? "green" : "red" }}> {item[1].symbol}</span>
                          
                            <span> {data.companyName}</span>

                            <span style={{ color : data.latestPrice < item[1].price ? "green" : "red" }}> ${item[1].price} </span>
                            
                          </div>

                        <div>
                          <span>{data.change}</span>
                          <span>{data.changePercent}%</span>
                        </div>
                      </div>
                     
                    }
                  </div>
                  )
                }
              )
            }
          </div>

        </div>

          {
            isModalShowing && 

            <div className="modal-container" onClick={()=>this.setState({isModalShowing: false})}>

              <div onClick={ (e)=> e.stopPropagation()} >
                
                    <AddSymbolBox exit={this.closeModal}/>
              </div>
            </div>
          }

        </div>
      );
    };
}

const mapStateToProps = ({ User_state, Portfolio_state }) => {

  return {
    token: User_state.token,
    portfolio : Portfolio_state.portfolio,
    portfolioSize: Portfolio_state.portfolio.length
  };
};

export default connect(
  mapStateToProps,
  null
)(AllWatchList);