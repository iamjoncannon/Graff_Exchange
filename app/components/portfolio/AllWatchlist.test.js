import React from 'react';
import renderer from 'react-test-renderer';
import { Provider } from 'react-redux';
import AllWatchlist from './AllWatchlist'
import configureStore from 'redux-mock-store';
import thunk from "redux-thunk"
const mockStore = configureStore([thunk]);
import { BrowserRouter as Router } from "react-router-dom";
import nocked from "../../../test/nock.setup.js"
import { connect } from 'tls';

const fake_api_data = {"data":
    {"hydrate_time_series_data":
        [{
            "date":"2019-10-02",
            "open":1206.42,
            "high":1206.42,
            "low":1173.37,
            "close":1176.985,
            "__typename":"Time_Series_Data_Point"
        }]
    }
}

const fake_store_with_data =  {
    User_state: {
        email: "yada@yada.com"
    },
    Portfolio_state: {
        token_error: "this is a token error",
        portfolio: {
            FB: {
                symbol: "FB",
                price: 100,
                current_holding: 5,
                historical: fake_api_data.data.hydrate_time_series_data,
                data: {
                    changePercent: .03,
                    latestPrice: 200
                }
            }
        },
        selectedPortfolioItem: "FB"
    }
}

const store_setup = (store) => {

    return renderer.create(
        <Provider store={store} >
           <Router >
               <AllWatchlist location={{pathname: "/"}}/>
           </Router>
       </Provider>
   );
}

describe("AllWatchlist", ()=>{

    let component 
    let store 
    let connected_react_component

    beforeEach(()=>{

        localStorage.setItem('yada@yada.com-hideList', '["GOOGL"]' ) 
                
        store = mockStore( fake_store_with_data );
    
        component = store_setup(store)
        connected_react_component = component.root.children[0].children[0].children[0].children[0]._fiber.stateNode
    })

    afterEach(()=>{

        store.clearActions();
    })

    it('1) renders successfully with data from redux store', ()=>{

        let tree = component.toJSON();
        
        expect(tree).toMatchSnapshot();
    })

    it('2) if portfolio is defined, maps over array and generates "watchlist-item-container" div for each', ()=>{
        
        const className = "watchlist-item-container"

        const divs = component.root.findAll(node => node.props.className === className)
    
        expect(divs).toBeTruthy()

        expect(divs.length).toEqual(1)
    })

    it('3) if data is defined, renders "watchlist-item" div with six spans', ()=>{
        
        const className = "watchlist-item"

        const parent = component.root.find(node => node.props.className === className)
    
        expect(parent).toBeTruthy()

        const spans = parent.props.children[0].props.children 

        expect(spans.length).toEqual(4)
    })
    
    it("4) if this.state.edit each portfolio item has edit icon on side", ()=>{

        renderer.act(()=>{

            connected_react_component.setState({edit: true})
        })

        const edit_icon = component.root.find(node => node.type === "i" && node.props.className.includes("plus") )

        expect(edit_icon).toBeTruthy()
    })
    
    it("5) clicking icon adds item symbol to hidelist", ()=>{
        
        renderer.act(()=>{

            connected_react_component.setState({edit: true})
        })

        const parent = component.root.find(node => node.props.className === "watchlist-item-container" )

        const icon = parent.props.children[0]

        renderer.act(()=>{

            icon.props.onClick()
        })

       expect(connected_react_component.state.hideList.includes("FB")).toBeTruthy()
    
    })
    
    it("6) handleEdit sets new hideList in local storage and sets state with new hidelist ", ()=>{

        renderer.act(()=>{

            connected_react_component.handleEdit(["", {symbol: "FB"}])
        })

        expect(localStorage).toEqual({ 'yada@yada.com-hideList': "[\"GOOGL\",\"FB\"]" })
    })
    
    it("7) when component mounts, sets state.hidelist to user's hide list in local storage ", ()=>{

        expect( localStorage.getItem('yada@yada.com-hideList') ).toEqual("[\"GOOGL\"]")
        
        expect(connected_react_component.state.hideList).toEqual(["GOOGL"])
    })
    
    it("8) if modal is showing, renders AddSymbolBox ", ()=>{

        renderer.act(()=>{

            connected_react_component.setState({isModalShowing: true})
        })       

        let box = component.root.find(node => node.type.name === "AddSymbolBox")

        expect(box).toBeTruthy()
    })
    
    it("9) if AddSymbolBox exit prop is called, closes modal box ", ()=>{

        renderer.act(()=>{

            connected_react_component.setState({isModalShowing: true})
        })

        let box = component.root.find(node => node.type.name === "AddSymbolBox")

        renderer.act(()=>{

            box.props.exit()
        })

        expect(connected_react_component.state.isModalShowing).toEqual(false)
    })
    
    it("10) if token-error, renders token error messages", ()=>{

        const errors = component.root.findAll(node => node.props.className === "error_message")
        
        expect(errors).toBeTruthy()
        expect(errors.length).toEqual(2)      
    })
})