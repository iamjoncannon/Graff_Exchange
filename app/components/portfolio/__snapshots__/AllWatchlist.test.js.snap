// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AllWatchlist 1) renders successfully with data from redux store 1`] = `
<div>
  <div
    className="all-watchlist"
  >
    <div>
      <i
        className="fas fa-plus fa-plus-edit fa-7x"
        onClick={[Function]}
      />
      <i
        className="fas fa-edit fa-7x"
        onClick={[Function]}
      />
    </div>
    <span>
      Portfolio
    </span>
    <div>
      <div>
        <span
          className="error_message"
        >
          Error validating account- please try signing in again.
        </span>
        <span
          className="error_message"
        >
          We apologize for the inconvenience
        </span>
      </div>
      <div
        className="watchlist-item-container"
      >
        <div
          className="watchlist-item"
          style={
            Object {
              "opacity": "1",
            }
          }
        >
          <div>
            <span
              style={
                Object {
                  "color": "red",
                }
              }
            >
               
              FB
            </span>
            <span>
               
            </span>
            $
            <span
              style={
                Object {
                  "color": "red",
                }
              }
            >
              100
            </span>
          </div>
          <div>
            <span />
            <span>
              0.03
              %
            </span>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
